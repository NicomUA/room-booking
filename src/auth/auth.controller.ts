import { Body, Controller, Post } from '@nestjs/common';
import { AuthService } from './auth.service';
import { LoginDto } from './dto/login.dto';
import { JwtPayloadResponse } from './dto/payload.dto';

@Controller()
export class AuthController {
  constructor(private readonly authService: AuthService) {}

  @Post('login')
  async login(
    @Body() { email, password }: LoginDto,
  ): Promise<JwtPayloadResponse> {
    return this.authService.login({ email, password });
  }
}
