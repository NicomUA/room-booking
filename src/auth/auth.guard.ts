import { ConfigurationService } from '@app/configuration';
import {
  CanActivate,
  ExecutionContext,
  Injectable,
  UnauthorizedException,
} from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { Request } from 'express';
import { JWTPayload } from './dto/payload.dto';

@Injectable()
export class AuthGuard implements CanActivate {
  constructor(
    private jwtService: JwtService,
    private config: ConfigurationService,
  ) {}

  async canActivate(context: ExecutionContext): Promise<boolean> {
    const request = context.switchToHttp().getRequest();
    const token = this.extractTokenFromHeader(request);
    const JWT_SECRET = this.config.get('JWT_SECRET');

    if (!token) {
      throw new UnauthorizedException('Token is missing');
    }
    try {
      const payload = await this.jwtService.verifyAsync<JWTPayload>(token, {
        secret: JWT_SECRET,
      });
      request['user'] = { id: payload.sub, ...payload };
    } catch {
      throw new UnauthorizedException();
    }
    return true;
  }

  private extractTokenFromHeader(request: Request): string | undefined {
    const [type, token] = request.headers.authorization?.split(' ') ?? [];
    return type === 'Bearer' ? token : undefined;
  }
}
