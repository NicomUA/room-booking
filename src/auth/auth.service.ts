import { DbService } from '@app/db';
import {
  BadRequestException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { LoginDto } from './dto/login.dto';
import { JwtPayloadResponse } from './dto/payload.dto';

@Injectable()
export class AuthService {
  constructor(
    private db: DbService,
    private jwtService: JwtService,
  ) {}

  async login({ email, password }: LoginDto): Promise<JwtPayloadResponse> {
    const user = await this.db.user.findUnique({
      where: { email },
    });

    if (!user) {
      throw new NotFoundException('User not found');
    }

    if (user.password !== password) {
      throw new BadRequestException('Wrong password');
    }

    const payload = { sub: user.id, email: user.email, name: user.name };
    return {
      access_token: await this.jwtService.signAsync(payload),
    };
  }
}
