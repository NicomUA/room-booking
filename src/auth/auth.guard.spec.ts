import { AuthGuard } from './auth.guard';
import { TestBed } from '@automock/jest';
import { JwtService } from '@nestjs/jwt';

describe('AuthGuard', () => {
  let guard: AuthGuard;
  let jwtService: jest.Mocked<JwtService>;

  const mockRequest = {
    originalUrl: '/',
    method: 'GET',
    params: undefined,
    query: undefined,
    body: undefined,
    headers: {
      authorization: 'Bearer test',
    },
  };

  const getRequestStub = jest.fn(() => mockRequest);

  const contextMock = {
    switchToHttp: jest.fn(() => ({
      getRequest: getRequestStub,
      getResponse: () => ({
        statusCode: 200,
      }),
    })),
    // method I needed recently so I figured I'd add it in
    getType: jest.fn(() => 'http'),
  };

  beforeAll(() => {
    const { unit, unitRef } = TestBed.create(AuthGuard).compile();
    guard = unit;
    jwtService = unitRef.get(JwtService);
  });

  it('should be defined', () => {
    expect(true).toBe(true);
  });

  it('canActivate', async () => {
    jwtService.verifyAsync.mockResolvedValue({ sub: 1 });
    const canActivate = await guard.canActivate(contextMock as any);
    expect(canActivate).toBe(true);
  });

  it('canActivate fail when token is invalid', async () => {
    jwtService.verifyAsync.mockRejectedValue({});
    const canActivate = async () => await guard.canActivate(contextMock as any);
    expect(canActivate).rejects.toThrowErrorMatchingSnapshot();
  });

  it('canActivate fail when token not provided', async () => {
    getRequestStub.mockReturnValue({
      ...mockRequest,
      headers: {
        authorization: undefined,
      },
    } as any);

    const canActivate = async () => await guard.canActivate(contextMock as any);
    expect(canActivate).rejects.toThrowErrorMatchingSnapshot();
  });
});
