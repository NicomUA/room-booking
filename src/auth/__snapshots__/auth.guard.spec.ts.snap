// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthGuard canActivate fail when token is invalid 1`] = `"Unauthorized"`;

exports[`AuthGuard canActivate fail when token not provided 1`] = `"Token is missing"`;
